### DISCUSSION TITLE
Is there a way to mock virtual module (module that doesn't exist)?

### DISCUSSION THREAD

**Discussion by Ky6uk** [vitest-dev/vitest#2114] [Q&A] [answered]
My question is about dynamic mocking modules that don't exist somewhere in the filesystem, for example dynamically generated on the build process after testing. Jest has [a parameter](https://jestjs.io/docs/jest-object#jestmockmodulename-factory-options:~:text=The%20third%20argument%20can%20be%20used%20to%20create%20virtual%20mocks%20%E2%80%93%20mocks%20of%20modules%20that%20don%27t%20exist%20anywhere%20in%20the%20system%3A) for such cases.

Not sure it's even possible, but I'm currently looking for a way to mock dynamic modules.

### Example

```ts
// file.ts

export async function loadLanguage(lang: string) : Promise<void> {
  // "../languages/" doesn't exist during testing
  const messages = await import(`../languages/${lang}.json`);

  setMessages(messages);
}
```

```ts
// file.test.ts

test((lang) => {
  vi.mock(`../languages/${lang}.json`, ({ woof: 'Woof!' }));

  await loadLanguage(lang);

  expect(setMessages).toHaveBeenCalled();
})
```

**Comment by sheremet-va** [comment, marked as answer]
You don't need parameter, just call vi.doMock or vi.mock with a path and factory. Beware that vi.mock is hoisted, so you can't use any outside variables. Refer to docs for examples

**Reply by Ky6uk** [reply]
@sheremet-va do you mean to mock just `'../languages'` directory? In the example above `setMessages` will not be called because `json` file doesn't exist and an error will be thrown. Mocking all possible variations like below also doesn't work.

```ts
vi.mock(`../languages/en.json`, ({ woof: 'Woof!' }));
vi.mock(`../languages/de.json`, ({ woof: 'Woof!' }));
```

**Reply by Ky6uk** [reply]
Also, in the example above the `loadLanguage` function that uses dynamic import might be handled and transformed by Vite's `dynamic-import-helper` somehow I'm probably not clearly understanding and maybe it requires mocking such a case somehow else.

**Reply by sheremet-va** [reply]
dynamic helper is only added, if it can parse the expression. To bypass that put import path into a variable:

```ts
const path = `../languages/${lang}.json`
const messages = await import(path)
```

Working mocking example: https://stackblitz.com/edit/vitest-dev-vitest-whfxjk?file=src.ts,test%2Fbasic.test.ts&initialPath=__vitest__

**Reply by Ky6uk** [reply]
In this case, Vite will not analyze this import and will not build related `.js` files after the build process if language files exist.

```
The above dynamic import cannot be analyzed by Vite.
See https://github.com/rollup/plugins/tree/master/packages/dynamic-import-vars#limitations for supported dynamic import formats. If this is intended to be left as-is, you can use the /* @vite-ignore */ comment inside the import() call to suppress this warning.
```

**Reply by sheremet-va** [reply]
Then you can mock `/@vite/dynamic-import-helper` instead of putting path into a variable:

```ts
vi.mock('/@vite/dynamic-import-helper', () => ({
  default: (_, path) => import(path)
})
```

**Reply by Ky6uk** [reply]
Ahh, that's a good idea, probably. I was pretty focused on case how to mock dynamic imports for inexisted modules and haven't thought out of this scope. Thanks.
